From c4a143ffd2611292639c238f0f57acd5820cda69 Mon Sep 17 00:00:00 2001
From: Tianshu Qiu <tian.shu.qiu@intel.com>
Date: Wed, 16 May 2018 11:19:56 +0800
Subject: [PATCH 058/129] media: intel-ipu4: imx319: Fix pm usage and add error
 handling

Message for Open Source:
Fix pm usage and add error handling.

Message for Internal:
[Issue/Feature] Sensor power control
[Root Cause/Changes] Fix pm usage and add error handling.

Change-Id: I81d886c4616a28f08a3ab720a93d0ea729b7f8fc
Tracked-On: #JICCCN-171
Signed-off-by: Tianshu Qiu <tian.shu.qiu@intel.com>
Signed-off-by: Meng Wei <wei.meng@intel.com>
---
 drivers/media/i2c/imx319.c | 18 +++++++-----------
 1 file changed, 7 insertions(+), 11 deletions(-)

diff --git a/drivers/media/i2c/imx319.c b/drivers/media/i2c/imx319.c
index 2d15e19..bb84981 100644
--- a/drivers/media/i2c/imx319.c
+++ b/drivers/media/i2c/imx319.c
@@ -1878,7 +1878,7 @@ static int imx319_set_ctrl(struct v4l2_ctrl *ctrl)
 	 * Applying V4L2 control value only happens
 	 * when power is up for streaming
 	 */
-	if (pm_runtime_get_if_in_use(&client->dev) <= 0)
+	if (pm_runtime_get_if_in_use(&client->dev) == 0)
 		return 0;
 
 	switch (ctrl->id) {
@@ -2253,7 +2253,8 @@ static int imx319_init_controls(struct imx319 *imx319)
 				0,
 				0,
 				link_freq_menu_items);
-	imx319->link_freq->flags |= V4L2_CTRL_FLAG_READ_ONLY;
+	if (imx319->link_freq)
+		imx319->link_freq->flags |= V4L2_CTRL_FLAG_READ_ONLY;
 
 	/* pixel_rate = link_freq * 2 * nr_of_lanes / bits_per_sample */
 	pixel_rate = (link_freq_menu_items[0] * 2 * 4) / 10;
@@ -2276,7 +2277,8 @@ static int imx319_init_controls(struct imx319 *imx319)
 	imx319->hblank = v4l2_ctrl_new_std(
 				ctrl_hdlr, &imx319_ctrl_ops, V4L2_CID_HBLANK,
 				hblank, hblank, 1, hblank);
-	imx319->hblank->flags |= V4L2_CTRL_FLAG_READ_ONLY;
+	if (imx319->hblank)
+		imx319->hblank->flags |= V4L2_CTRL_FLAG_READ_ONLY;
 
 	exposure_max = mode->fll_def - 18;
 	imx319->exposure = v4l2_ctrl_new_std(
@@ -2381,10 +2383,9 @@ static int imx319_probe(struct i2c_client *client,
 	 * Device is already turned on by i2c-core with ACPI domain PM.
 	 * Enable runtime PM and turn off the device.
 	 */
-	pm_runtime_get_noresume(&client->dev);
 	pm_runtime_set_active(&client->dev);
 	pm_runtime_enable(&client->dev);
-	pm_runtime_put(&client->dev);
+	pm_runtime_idle(&client->dev);
 
 	return 0;
 
@@ -2409,14 +2410,9 @@ static int imx319_remove(struct i2c_client *client)
 	media_entity_cleanup(&sd->entity);
 	imx319_free_controls(imx319);
 
-	/*
-	 * Disable runtime PM but keep the device turned on.
-	 * i2c-core with ACPI domain PM will turn off the device.
-	 */
-	pm_runtime_get_sync(&client->dev);
 	pm_runtime_disable(&client->dev);
 	pm_runtime_set_suspended(&client->dev);
-	pm_runtime_put_noidle(&client->dev);
+
 	mutex_destroy(&imx319->mutex);
 
 	return 0;
-- 
1.9.1

